source/engine/graphics2/diligent/DiligentEngineLinux.cpp

```diff
@@ -0,0 +1,112 @@
+    auto preferredApiOrder = std::vector<std::string_view>{};
+    preferredApiOrder.reserve(2);
+    std::ranges::copy(supportedApis, std::back_inserter(preferredApiOrder));
+    RotateElementToBeginning(preferredApiOrder, renderApi);
+
+    std::string errorMessage;
+    LinuxNativeWindow window;
+    window.WindowId = static_cast<Diligent::Uint32>(glfwGetX11Window(window_.GetWindowHandle()));
+    window.pDisplay = glfwGetX11Display();
+    SwapChainDesc SCDesc;
+
+    /* Initialize the device and context. First try to init the preferred API. Fall back to others on failure. */
+    for (const auto& api : preferredApiOrder)
+    {
+        if (api == api::Vulkan)
+        {
+            try
+            {
+                #if EXPLICITLY_LOAD_ENGINE_VK_DLL
+                    auto* GetEngineFactoryVk = LoadGraphicsEngineVk();
+                #endif
+
+                EngineVkCreateInfo engineCI;
+                auto* pFactoryVk = GetEngineFactoryVk();
+                pFactoryVk->CreateDeviceAndContextsVk(engineCI, &m_pDevice, &m_pImmediateContext);
+
+                if (!graphicsSettings.isHeadless)
+                    pFactoryVk->CreateSwapChainVk(m_pDevice, m_pImmediateContext, SCDesc, window, &m_pSwapChain);
+                    
+                m_renderApi = api;
+                NC_LOG_TRACE("Successfully initialized the Vulkan rendering engine.");
+                break;
+            }
+            catch (const std::runtime_error& e)
+            {
+                EnsureContextFlushed(m_pImmediateContext);
+                NC_LOG_WARNING("Failed to initialize Vulkan.");
+                errorMessage = fmt::format("{0} Failed to initialize Vulkan: {1} \n", errorMessage, e.what());
+            }
+        }
+        else if (api == api::OpenGL)
+        {
+            try
+            {
+                #if EXPLICITLY_LOAD_ENGINE_GL_DLL
+                    auto GetEngineFactoryOpenGL = LoadGraphicsEngineOpenGL();
+                #endif
+
+                auto* pFactoryOpenGL = GetEngineFactoryOpenGL();
+                EngineGLCreateInfo engineCI;
+                glfwMakeContextCurrent(window_.GetWindowHandle());
+                engineCI.Window = window;
+
+                if (!graphicsSettings.isHeadless)
+                    pFactoryOpenGL->CreateDeviceAndSwapChainGL(engineCI, &m_pDevice, &m_pImmediateContext, SCDesc, &m_pSwapChain);
+                
+                m_renderApi = api;
+                NC_LOG_TRACE("Successfully initialized the OpenGL rendering engine.");
+                break;
+            }
+            catch (const std::runtime_error& e)
+            {
+                EnsureContextFlushed(m_pImmediateContext);
+                NC_LOG_WARNING("Failed to initialize OpenGL.");
+                errorMessage = fmt::format("{0} Failed to initialize OpenGL: {1} \n", errorMessage, e.what());
+            }
+        }
+    }
+
+    if (m_renderApi == "" || errorMessage != "")
```

Comment:
If vulkan failed but falling back to opengl succeeded, won't this throw because errorMessage isn't empty?

Reply:
Yes - great catch!